### PR TITLE
Support using the test file system in SwiftDocC tests

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: #663 

## Summary

This is a step towards using `TestFileSystem` in more tests, allowing it to be used in SwiftDocCTests. Previously it was only available in SwiftDocC**Utilities**Tests.

It then makes `DiagnosticConsoleWriter` support the test file system (#663) and updates one test to use it.

Lastly, it updates how `TestFileSystem` provides bundles to behave like `LocalFileSystemDataProvider` and removes a lot of unintentional printing in tests.

> Note:
> There's a lot more we _could_ add to the new common target but I feel that we should move slow and deliberately on that and only move files as an action towards other goals (for example making it easy to write good tests).

## Dependencies

None.

## Testing

Nothing in particular. This shouldn't have changed any functionality.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [ ] Updated documentation if necessary

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocCTestUtilities/FilesAndFolders.swift [hunk 1]
```diff
@@ -93,51 +93,41 @@ public struct InfoPlist: File, DataRepresentable {
...
             case versionString = "CFBundleVersion"
-            case developmentRegion = "CFBundleDevelopmentRegion"
         }
     }
 
     public func data() throws -> Data {
-        // TODO: Replace this with PropertListEncoder (see below) when it's available in swift-corelibs-foundation
-        // https://github.com/apple/swift-corelibs-foundation/commit/d2d72f88d93f7645b94c21af88a7c9f69c979e4f
-        let infoPlist = [
+        let encoder = PropertyListEncoder()
```
Comment (patshaughnessy) [inline Sources/SwiftDocCTestUtilities/FilesAndFolders.swift:123]: Another unrelated fix?

File: Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift [hunk 2]
```diff
@@ -44,8 +44,7 @@ class ConvertActionStaticHostableTests: StaticHostingBaseTests {
             hostingBasePath: basePath,
             temporaryDirectory: try createTemporaryDirectory()
         )
-       
-        _ = try action.perform(logHandle: .standardOutput)
+        _ = try action.perform(logHandle: .none)
```
Comment (patshaughnessy) [inline Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift:47]: Why bother avoiding stdout during test runs? Is there any harm in emitted something to stdout for test runs? Doesn't `swift test` automatically filter and/or discard anything tests write to stdout?

**Review by patshaughnessy — APPROVED** [review]
Looks like a useful enhancement 👏 

But why bother creating a new target ("_Common")? Seems like that just increases the cognitive load and complexity of DocC without much benefit. Couldn't you just add `TestFileSystem` or whichever new classes you need to the normal target? And use them in the test code?

File: Sources/SwiftDocCTestUtilities/FilesAndFolders.swift [hunk 3]
```diff
@@ -93,51 +93,41 @@ public struct InfoPlist: File, DataRepresentable {
     /// The information that the Into.plist file contains.
     public let content: Content
 
-    public init(displayName: String, identifier: String, versionString: String = "1.0", developmentRegion: String = "en") {
+    public init(displayName: String, identifier: String? = nil, versionString: String = "1.0") {
```
Comment (daniel-grumberg) [inline Sources/SwiftDocCTestUtilities/FilesAndFolders.swift:96]: I am not familiar with this, but what was "developmentRegion" and why was it removed?

File: Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift (hunk 2 above)
Comment (daniel-grumberg) [reply Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift:47]: I am actually in favor of this, no need to write to stdout and potentially slow down test execution due to console writing, additionally in parallel test runs this ends up being a bit garbled anyway.

**Comment by daniel-grumberg** [comment]
This looks like a great improvements to our tests. I wonder if we can come up with a better module name than _Common especially since it does include public API.

**Comment by d-ronnqvist** [comment]
> This looks like a great improvements to our tests. I wonder if we can come up with a better module name than _Common especially since it does include public API.

The idea is for this target to _not_ be public API. When we update to Swift 5.9 we can use the new `package` access level everywhere in this module. Unfortunately it's not possible to hide the target, so other projects can still import it but wouldn't be able to access any API. I used the underscore prefix to signal to other developers that this target isn't intended to be public.

File: Sources/SwiftDocCTestUtilities/FilesAndFolders.swift (hunk 3 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocCTestUtilities/FilesAndFolders.swift:96]: It's a configuration that we never parsed and never did anything with so there wasn't any point in writing it to the Info.plist in all the tests.

File: Sources/SwiftDocCTestUtilities/FilesAndFolders.swift (hunk 1 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocCTestUtilities/FilesAndFolders.swift:123]: Yes. I encountered this TODO comment and saw that we could do it now.

File: Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift [hunk 4]
```diff
@@ -27,18 +28,21 @@ public final class DiagnosticConsoleWriter: DiagnosticFormattingConsumer {
...
     ///   - baseUrl: A url to be used as a base url when formatting diagnostic source path.
     ///   - highlight: Whether or not to highlight the default diagnostic formatting output.
+    ///   - fileManager: The file manager that reads the contents of source files.
     public init(
         _ stream: TextOutputStream = LogHandle.standardError,
         formattingOptions options: DiagnosticFormattingOptions = [],
         baseURL: URL? = nil,
-        highlight: Bool? = nil
+        highlight: Bool? = nil,
+        fileManager: FileManagerProtocol = FileManager.default
```
Comment (daniel-grumberg) [inline Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:37]: If we don't want to expose _Common as public API then we need to figure out another strategy for this initializer.

Comment (daniel-grumberg) [reply Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:37]: Essentially if we don't want to expose _Common in our API surface then we need to make sure that it isn't used in any public API declarations. I am not sure but I think that FileManagerProtocol is potentially a useful thing to expose though, so maybe another strategy is required.

File: Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift (hunk 2 above)
Comment (d-ronnqvist) [reply Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift:47]: The writes to stdout and stderr is only filtered out when running tests in parallel, for example using `swift test --parallel`. My main reason for removing it in the tests is to reduce noise for local development workflows when running a single test repeatedly.

File: Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift [hunk 5]
```diff
@@ -360,7 +368,7 @@ extension DefaultDiagnosticConsoleFormatter {
             // Example:
             // 9  | A line outside the diagnostic range.
             // 10 + A line inside the diagnostic range.
-            result.append("\n\(linePrefix) \(separator) \(highlightedSource)")
+            result.append("\n\(linePrefix) \(separator) \(highlightedSource)".removingTrailingWhitespace())
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:371]: Yes. I was updating a test diagnostics formatter [here](https://github.com/apple/swift-docc/pull/756/files#diff-856201a2266867cc3bce33c8e301d89e125a957bd4f7e0d834abd45f5fd4c561R249) to use the test file system and I wanted to assert that the diagnostic message was logged. 

The reason for this change is that the two empty lines of source in the diagnostic message in that test was really annoying to work with because the console writer didn't trim the trailing whitespace, so it printed those two lines as `2 | ` and `4 | ` (with a trailing space). The issue is that Xcode automatically trims trailing whitespace while editing so I couldn't write the string literal for the expected diagnostic message because Xcode would remove the trailing whitespace and make the test fail.

I could disable that setting in Xcode, write the string literal, and then enable that setting again, but anyone else who touched that test in the future would break it. 

I could also write the test so that it splits the diagnostic message by line, trims the trailing whitespace, and joins the diagnostic message again before comparing it to the expected value, but we would have to repeat that code in every test that wanted to assert that a diagnostic message was printed as expected.

So instead I decided to trim the trailing whitespace in the diagnostic message so that it will compare nicely to the expected output in this test and in future tests.

File: Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift (hunk 2 above)
Comment (patshaughnessy) [reply Tests/SwiftDocCUtilitiesTests/ConvertActionStaticHostableTests.swift:47]: Sounds great thank you. I was just curious...

**Comment by d-ronnqvist** [comment]
> But why bother creating a new target ("_Common")? Seems like that just increases the cognitive load and complexity of DocC without much benefit. Couldn't you just add `TestFileSystem` or whichever new classes you need to the normal target? And use them in the test code?

We have two targets (one for the central logic and one for the command line structure) and only _one_ of those targets has access to `FileManagerProtocol` and `TestFileSystem`. We've already encountered a few situations where tests for DocC need to write real files on the file system.

File: Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift (hunk 4 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Diagnostics/DiagnosticConsoleWriter.swift:46]: Hm... it may be best to discuss what direction we want to go in with the team. 

I did avoid exposing _Common in https://github.com/apple/swift-docc/pull/756/commits/4fc6a1644e06eab3fd723b63eda9fb7dac5d1dcf so that we can change `FileManagerProtocol` from `public` to `package` in the future without breaking any public API but we would need to be careful so that we don't introduce new breaking points accidentally. 

As an alternative we could bump the Swift version to 5.9 now and make `FileManagerProtocol` package accessible. We can always increase the visibility in the future bug going is a lot more time and work.

**Comment by d-ronnqvist** [comment]
We discussed this within the team and decided to make a smaller change to the project structure for this PR. 

Since SwiftDocCUtilities already depends on SwiftDocC, I will move FileManagerProtocol into SwiftDocC and make it SPI so that both targets can use it that way. In the future when we are ready to update to Swift 5.9 we may change that from SPI to "package" access level.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
> We discussed this within the team and decided to make a smaller change to the project structure for this PR.
> 
> Since SwiftDocCUtilities already depends on SwiftDocC, I will move FileManagerProtocol into SwiftDocC and make it SPI so that both targets can use it that way. In the future when we are ready to update to Swift 5.9 we may change that from SPI to "package" access level.

I made these changes in https://github.com/apple/swift-docc/pull/756/commits/9068742e95df38bd1b69cfc1f141ed9144896847

Until we update to Swift 5.9 we'll need to do a use `@_spi(FileManagerProtocol) import SwiftDocC` in new places but the benefit is that we can't accidentally use `FileManagerProtocol` in our public API.

**Review by daniel-grumberg — APPROVED** [review]
